src/compiler/checker.ts

```diff
@@ -5595,9 +5595,14 @@ namespace ts {
 
                 if (source.flags & TypeFlags.TypeParameter) {
                     let constraint = getConstraintOfTypeParameter(<TypeParameter>source);
+
                     if (!constraint || constraint.flags & TypeFlags.Any) {
                         constraint = emptyObjectType;
                     }
+
+                    // The constraint may need to be further instantiated with its 'this' type.
+                    constraint = getTypeWithThisArgument(constraint);

```

Comment:
Yup. I could have introduced an `else` branch, but I figured that this would be a little simpler to read. I'm open to changing it.